lib/m2ts/caption-stream.js

```diff
@@ -319,77 +494,162 @@ var Cea608Stream = function() {
       this.startPts_ = packet.pts;
       break;
 
-    case ROLL_UP_2_ROWS:
+    case (this.control_ | ROLL_UP_2_ROWS):
       this.topRow_ = BOTTOM_ROW - 1;
       this.mode_ = 'rollUp';
       break;
-    case ROLL_UP_3_ROWS:
+    case (this.control_ | ROLL_UP_3_ROWS):
       this.topRow_ = BOTTOM_ROW - 2;
       this.mode_ = 'rollUp';
       break;
-    case ROLL_UP_4_ROWS:
+    case (this.control_ | ROLL_UP_4_ROWS):
       this.topRow_ = BOTTOM_ROW - 3;
       this.mode_ = 'rollUp';
       break;
-    case CARRIAGE_RETURN:
+    case (this.control_ | CARRIAGE_RETURN):
+      this.clearFormatting(packet.pts);
       this.flushDisplayed(packet.pts);
       this.shiftRowsUp_();
       this.startPts_ = packet.pts;
       break;
 
-    case BACKSPACE:
+    case (this.control_ | BACKSPACE):
       if (this.mode_ === 'popOn') {
         this.nonDisplayed_[BOTTOM_ROW] = this.nonDisplayed_[BOTTOM_ROW].slice(0, -1);
       } else {
         this.displayed_[BOTTOM_ROW] = this.displayed_[BOTTOM_ROW].slice(0, -1);
       }
       break;
-    case ERASE_DISPLAYED_MEMORY:
+    case (this.control_ | ERASE_DISPLAYED_MEMORY):
       this.flushDisplayed(packet.pts);
       this.displayed_ = createDisplayBuffer();
       break;
-    case ERASE_NON_DISPLAYED_MEMORY:
+    case (this.control_ | ERASE_NON_DISPLAYED_MEMORY):
       this.nonDisplayed_ = createDisplayBuffer();
       break;
+
     default:
-      char0 = data >>> 8;
       char1 = data & 0xff;
 
-      // Look for a Channel 1 Preamble Address Code
-      if (char0 >= 0x10 && char0 <= 0x17 &&
-          char1 >= 0x40 && char1 <= 0x7F &&
-          (char0 !== 0x10 || char1 < 0x60)) {
-        // Follow Safari's lead and replace the PAC with a space
-        char0 = 0x20;
-        // we only want one space so make the second character null
-        // which will get become '' in getCharFromCode
-        char1 = null;
-      }
+      var text = '';
+
+      // Handle special characters
+      if (char0 === this.ext_ && char1 >= 0x30 && char1 <= 0x3f) {
+        // Bitmask char0 so that we can apply character transformations
+        // regardless of field and data channel.
+        // Then byte-shift to the left and OR with char1 so we can pass the
+        // entire character code to `getCharFromCode`.
+        char0 = (char0 & 0x03) << 8;
+        text = getCharFromCode(char0 | char1);
+        this.column_++;
+
+      // Handle extended characters
+      } else if ((char0 === (this.ext_ + 1) || char0 === (this.ext_ + 2)) &&
```

Comment:
Hmmmmm.......... While I agree with you that `char0 === (this.ext_ + 1)` isn't easy to read or understand, I don't particularly like `this.extSetSpanishFrenchMisc_ = 0x13` either. Someone unfamiliar with the spec might mistakenly understand that to mean that `0x12` indicates an extended character, which is not the case -- it can also be a PAC. This is even more confusing for `0x11`, which can be a special character, a mid-row code, or a PAC.

If you think the comments aren't enough, how do you feel about object methods? Something like
```javascript
if (this.isSpecialCharacter(data)) {
  // do the thing
} else if (this.isExtCharacter(data)) {
  // do the other thing
} // etc...

/**
 * Detects if the 2-byte packet data is a special character
 *
 * Special characters have a second byte in the range 0x30 to 0x3f,
 * with the first byte being 0x11 (for data channel 1) or 0x19 (for
 * data channel 2).
 *
 * @param  {Integer} data  The two cc_data bytes
 * @return {Boolean}       Whether the data is an special character
 */
Cea608Stream.prototype.isSpecialCharacter = function(char0, char1) {
  return (char0 === this.ext_ && char1 >= 0x30 && char1 <= 0x3f);
}

/**
 * Detects if the 2-byte packet data is an extended character
 *
 * Extended characters have a second byte in the range 0x20 to 0x3f,
 * with the first byte being 0x12 or 0x13 (for data channel 1) or
 * 0x1a or 0x1b (for data channel 2).
 *
 * @param  {Integer} data  The two cc_data bytes
 * @return {Boolean}       Whether the data is an extended character
 */
Cea608Stream.prototype.isExtCharacter = function(char0, char1) {
  return ((char0 === (this.ext_ + 1) || char0 === (this.ext_ + 2)) &&
    (char1 >= 0x20 && char1 <= 0x3f));
}
```

That way the general flow is nice and abstracted, but the gory implementation bits are in the corresponding function.